pkg/nvml/device.go

```diff
@@ -15,9 +15,38 @@
 package nvml
 
 import (
+	"fmt"
+	"reflect"
 	"unsafe"
 )
 
+// nvmlDeviceHandle attempts to convert a device d to an nvmlDevice.
+// This is required for functions such as GetTopologyCommonAncestor which
+// accept Device arguments that need to be passed to internal nvml* functions
+// as nvmlDevice parameters.
+func nvmlDeviceHandle(d Device) nvmlDevice {
+	if nvmlDevice, ok := d.(nvmlDevice); ok {
+		return nvmlDevice
+	}
+
+	val := reflect.ValueOf(d)
+	typ := val.Type()
+
+	if typ.Kind() == reflect.Pointer {
```

Comment:
Is required for:
```
panic: reflect: NumField of non-struct type *device.device
```

Reply:
Instead of this, can we replace `val := reflect.ValueOf(d)` above with:
```
	val := reflect.ValueOf(d)
	for val.Kind() == reflect.Ptr {
		val = val.Elem()
	}
```

Reply:
Do we also need to check to see if it is a struct type and panic with a more convenient error if it is not?

Reply:
I added an explicit check. Not sure if we add more value, but at least we have a point where we can extend the type checks.

As a matter of interest, why did you use `for` above instead of `if`?